staging/src/k8s.io/apiextensions-apiserver/test/integration/fieldselector_test.go

```diff
@@ -0,0 +1,868 @@
+		select {
+		case <-upCh:
+			return true, nil
+		default:
+			t.Logf("Waiting for webhook to become effective, getting marker object: %v", err)
+			return false, nil
+		}
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	var tcs []selectableFieldTestCase
+	for _, version := range []string{"v1", "v1beta1"} {
+		var colorSelector string
+		switch version {
+		case "v1":
+			colorSelector = "spec.color"
+		case "v1beta1":
+			colorSelector = "spec.hue"
+		}
+
+		tcs = append(tcs, []selectableFieldTestCase{
+			{
+				version:        version,
+				fieldSelector:  fmt.Sprintf("%s=blue", colorSelector),
+				expectedByName: sets.New("shirt1", "shirt2"),
+			},
+			{
+				version:        version,
+				fieldSelector:  "spec.quantity=2",
+				expectedByName: sets.New("shirt1", "shirt3"),
+			},
+			{
+				version:        version,
+				fieldSelector:  "spec.size=M",
+				expectedByName: sets.New("shirt2"),
+			},
+			{
+				version:        version,
+				fieldSelector:  "spec.branded=false",
+				expectedByName: sets.New("shirt2", "shirt3"),
+			},
+			{
+				version:        version,
+				fieldSelector:  fmt.Sprintf("%s=blue,spec.quantity=2", colorSelector),
+				expectedByName: sets.New("shirt1"),
+			},
+			{
+				version:        version,
+				fieldSelector:  fmt.Sprintf("%s=blue,spec.branded=false", colorSelector),
+				expectedByName: sets.New("shirt2"),
+			},
+			{
+				version:        version,
+				fieldSelector:  "spec.nosuchfield=xyz",
+				expectedByName: sets.New[string](),
+				expectError:    "field label not supported: spec.nosuchfield",
+			},
+		}...)
+	}
+
+	t.Run("watch", func(t *testing.T) {
+		testWatch(ctx, t, tcs, dynamicClient)
+	})
+	t.Run("list", func(t *testing.T) {
+		testList(ctx, t, tcs, dynamicClient)
+	})
+	t.Run("deleteCollection", func(t *testing.T) {
+		testDeleteCollection(ctx, t, tcs, dynamicClient)
+	})
+}
+
+func testWatch(ctx context.Context, t *testing.T, tcs []selectableFieldTestCase, dynamicClient dynamic.Interface) {
+	clients := map[string]dynamic.NamespaceableResourceInterface{}
+	for _, version := range []string{"v1", "v1beta1"} {
+		clients[version] = dynamicClient.Resource(schema.GroupVersionResource{Group: "tests.example.com", Version: version, Resource: "shirts"})
+	}
+
+	deleteTestResources(ctx, t, dynamicClient)
+	watches := map[string]watch.Interface{}
+	for _, tc := range tcs {
+		shirtClient := clients[tc.version]
+		w, err := shirtClient.Watch(ctx, metav1.ListOptions{FieldSelector: tc.fieldSelector})
+		if len(tc.expectError) > 0 {
+			if err == nil {
+				t.Errorf("Expected error but got none while creating watch for %s", tc.Name())
+			}
+			continue
+		}
+		if err != nil {
+			t.Fatalf("failed to create watch for %s: %v", tc.Name(), err)
+		} else {
+			watches[tc.Name()] = w
+		}
+	}
+	defer func() {
+		for _, w := range watches {
+			w.Stop()
+		}
+	}()
+
+	createTestResources(ctx, t, dynamicClient)
+
+	// after creating resources, delete one to make sure deletions can be observed
```

Comment:
Either in this test or a different one, can we change a color and see it disappear from the watch?